First pass at re-joining persisted rooms.

natesholland opened the pull request:
This is my first pass at re-joining persisted rooms, I think we need to move the persisted_rooms to not be private on Lita::Robot so that the adapter can access that list and re-join on start up. Once that is done I think we can start writing more tests around this and then it may be ready to merge.

natesholland commented:
This needs https://github.com/jimmycuadra/lita/pull/119 to be merged in first in order to make this feasible.

natesholland commented:
Ok I made it so that this now looks at the persisted rooms list in redis on start up and then loads them in. Also added specs around this to make sure it behaves as expected.

jimmycuadra commented:
It should use the IDs passed into the `:loaded` event rather than accessing Redis:

``` ruby
on :connected, :store_room_ids

def initialize(robot)
  # ...
  room_ids.each { |room_id| join(room_id) }
  # ...
end

def store_room_ids(payload)
  self.room_ids = Array(payload[:room_ids])
end

private

attr_accessor :room_ids
```

That line to loop through the rooms and join them already exists above the line you added to call the `join_persisted_rooms` method, so you can remove the new call.

jimmycuadra commented:
Also: Looks like you should rebase, since the travis.yml changes are showing up as part of this diff.

natesholland commented:
I'm getting pretty confused here as to why you want that specific implementation. I think it would be better to have Lita be in charge of connecting to the rooms on start up. In these implementations we have a weird coupling of the adapter to the robot and how persisted rooms are stored and joined. It seems to me like all of that logic should be kept in one place and not split up across gems.

Also when I try do do `on :loaded, :store_room_ids` I get an error for `undefined method `on' for Lita::Adapters::HipChat:Class (NoMethodError)`

I will keep looking at this but maybe it would help if you could explain why you want lita to be in charge of making the persisted rooms list but then the adapter be in charge of consuming it? Also do we want to trigger on `:loaded` or `:connected` and does it matter?

natesholland commented:
In order to handle triggers it looks like I will need to pull in `lita/handler/event_router` Do I want to be pulling logic from handlers into an adapter? This seems like I'm breaking a separation of concerns.

jimmycuadra commented:
Wow, you are right. I made a huge error in thinking that adapters could listen to events already. They really should be able to. I merged your PR making `Lita::Robot#persisted_rooms` public and released it in Lita 4.4.2, so please go ahead with your original plan. Sorry to lead you astray! :(

natesholland commented:
ok @jimmycuadra I moved back towards my original implementation with a few tweaks. I added some specs around this to make sure it all works and I think I'm finally getting close to a solid implementation. I will want to squash these before we merge though because there's a lot of flailing which I don't think needs to be in the main repo.

natesholland commented:
Also I threw the travis.yml stuff in there because I figured we needed it but if you would like I can pull it out into its own PR

jimmycuadra commented:
I am really off my game this week! I was confused with the Travis changes on the lita repo. Yes, it's cool to add them in here.

The only thing I'd change now is removing the rescue clause when joining rooms. I'd prefer to just let any exception be raised there. After that, squash your commits and I'll merge! Thanks so much.

natesholland commented:
I wish we didn't need the rescue there but here is the problem. Since the rooms entered into join are user generated we could easily run into a situation where a user said 'lita join bad_room' and then when lita starts back up lita will see 'bad_room' in persisted rooms and attempt to join, but since the room doesn't exist an error gets thrown. This keeps that from happening because Lita will just give up on malformed room names.

natesholland commented:
The reason we can get bad room names in the persisted rooms is the join method on the robot looks like this:

```ruby
    def join(room_id)
      Lita.redis.sadd("persisted_rooms", room_id)
      adapter.join(room_id)
    end
```

So you can see how a bad room name could make it into that persisted_rooms redis set and them the bot won't start back up because it will attempt to join a room that doesn't exist (or one that got deleted during reboot, or whatever) and then fail on start up.

natesholland commented:
@jimmycuadra if you agree with me on the rescue this is squashed into logical commits and ready to go :)

jimmycuadra commented:
Good catch. I'd say that's a bug in Lita's persisted rooms logic. It should verify that the room exists before attempting to join it. Go ahead and remove the rescue clause and I will release a bug fix for Lita that makes the join a no-op if it can't find a room for the given ID.

natesholland commented:
Ok @jimmycuadra all ready to get pulled in!

jimmycuadra commented:
Thank you, Nate!

natesholland commented:
Jimmy can you bump a gem version and release this please? I would like to get this pulled into the Lita bot here at Spiceworks

jimmycuadra commented:
Pretty soon! I need to do a little more before the next release. You can use a Git dependency in the Gemfile if you need the feature right away.
